### PR TITLE
Compatibility with mkosi v15+

### PR DESCRIPTION
**Pull request by marc-hb**
4 commits

### CODE REVIEW

**Comment by stellarhopper** [comment]
Doing some quick tests with this on Fedora 40 (mkosi v22) I noticed a few things:

1. It looks like the ESP partition (/efi) isn't mounted on the guest as it would be pre-v14. This is probably not critical - I tried with run_qemu's --no-direct-kernel, so it does seem to boot from the image instead of qemu's -kernel / -append options. The goal with this was to produce an image that could be booted on real hardware or a different emulator.
2. The default roof filesystem is now Fedora's default - btrfs. I think pre-v14's gpt_ext4 option for the output caused it to be ext4. I don't think this is a big deal either - users may have to enable btrfs in their kernel configs (I didn't have it on).
3. pre-v14, we used mkosi.postinst to build the 'ndctl' tree in the guest. See https://github.com/pmem/run_qemu/blob/main/run_qemu.sh#L919
This appears to be broken (I didn't see any ndctl build output in the mkosi build stages, nor any binaries in the guest. I'm guessing the postinst functionality has also been reworked/renamed by mkosi - it would be nice to get this functional again before making the switch.

**Comment by marc-hb** [comment]
Thanks @stellarhopper for all the testing!

>  it would be nice to get this functional again before making the switch.

Just to avoid any confusion: this PR does NOT make a "switch" to mkosi v15+. In theory, **support for mkosi v14- should be 100% unaffected**. Can you confirm?

The goal of this PR is to get "something" _partially_ working with mkosi v15+ for the very first time and you sound like this is achieved? I mean for you too - not just for me :-)

Another, very important goal is to get "something" working in CI too - also achieved with GitHub's Ubuntu 22.04 runner. This is a major milestone because "something" is infinitely better than "nothing" for catching regressions.

tl;dr: this is an important milestone but not the end of the mkosi v15+ journey.

**Comment by marc-hb** [comment]
> The default roof filesystem is now Fedora's default - btrfs. I think pre-v14's gpt_ext4 option for the output caused it to be ext4.

Yes: I didn't bother with (over?)specifying the disk format yet because it's a brand new configuration file (`systemd-part`) and I didn't look into it (yet). So, the preferred format is whatever the distro prefers and Fedora defaults to btrfs. Other distros might default to something else.

> pre-v14, we used mkosi.postinst to build the 'ndctl' tree in the guest.  I'm guessing the postinst functionality has also been reworked/renamed by mkosi

I will definitely take a look but I don't think this gap should block this PR.

> It looks like the ESP partition (/efi) isn't mounted on the guest as it would be pre-v14.

Mmmm... weird: as you can see from this and #73, I know the ESP is present in the image for sure.  It shows up in `/proc/partitions` after `losetup -Pf, otherwise `run_qemu.sh` fails, long before trying to boot.  If it's there and visible on the host, then why would it not appear in the guest? Again I'm only speaking about v15+; please confirm this PR makes no difference with v14-. That would definitely be a blocker!

**Comment by stellarhopper** [comment]
> Yes: I didn't bother with (over?)specifying the disk format yet because it's a brand new configuration file (systemd-part) and I didn't look into it (yet). So, the preferred format is whatever the distro prefers and Fedora defaults to btrfs. Other distros might default to something else.

Yep, that's reasonable to default to the distro's filesystem, unless there is any reason to deviate from that.

> Mmmm... weird: as you can see from this and #73, I know the ESP is present in the image for sure. It shows up in `/proc/partitions` after `losetup -Pf`, otherwise `run_qemu.sh` fails - long before even trying to boot. If the ESP exists and is visible on the host, then why would it not appear in the guest? Again I'm only speaking about v15+; please confirm this PR makes no difference with v14-. That would definitely be a blocker!

Correct, v14 still works the same as before, so this can go ahead.

**Comment by marc-hb** [comment]
> It looks like the ESP partition (/efi) isn't mounted on the guest as it would be pre-v14. 

I just tried with Fedora's `mkosi14` and the ESP partition is not mounted in the guest either.

> The goal with this was to produce an image that could be booted on real hardware or a different emulator.

The default qemu command line with `mkosi14` looks strange. One one hand it has this: `drive if=pflash,format=raw,unit=0,file=OVMF_CODE.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=OVMF_VARS.fd`

These look like `run_qemu` wants to make sure the disk image can be booted as you just wrote. 

But then it has this too: `-kernel mkosi.extra/boot/vmlinuz-6.12.0 -initrd mkosi.extra/boot/initramfs-6.12.0.img`

The latter makes sure the ability to boot the disk image is... NOT tested? Confusing.

**Comment by stellarhopper** [comment]
> I wonder why "mounted by default" matters to Linux BTW. The ESP partition is for UEFI/OVMF, not for Linux.

Right, it doesn't matter for Linux, it was just a way to make sure its contents are sane when preparing a disk image for use elsewhere.

> But then it has this too: `-kernel mkosi.extra/boot/vmlinuz-6.12.0 -initrd mkosi.extra/boot/initramfs-6.12.0.img`
> 
> The latter makes sure the ability to boot the disk image is... NOT tested? Confusing.

Those *shouldn't*be getting used if specifying run_qemu's --no-direct-kernel option.

**Comment by marc-hb** [comment]
> Those shouldn'tbe getting used if specifying run_qemu's --no-direct-kernel option.

Thanks, I didn't know about that option. I tried it with both `mkosi14` and `mkosi v22` on Fedora. It works with both: the ESP partition is successful with both. The ESP gets mounted by default in neither.

So long story short I have found zero difference yet in this area between `mkosi14` and `mkosi v22` on Fedora. If you can still see one, could you file a bug with very detailed versions and reproduction steps? 

I will look at `ndctl` next.

**Comment by marc-hb** [comment]
> So long story short I have found zero difference yet in this area between mkosi14 and mkosi v22 on Fedora.

Correct myself in new issue #75

**Comment by marc-hb** [comment]
> pre-v14, we used mkosi.postinst to build the 'ndctl' tree in the guest. I'm guessing the postinst functionality has also been reworked/renamed by mkosi



Confirmed:

https://github.com/systemd/mkosi/commit/9b626c647037bc8a9281da9a4a5c62624089183a "Run scripts on the host by default (v15)

There is a recommended migration path, looking into it.
